simulacrum/dataset.py

```diff
@@ -1,185 +1,97 @@
-            logging.error('cols and types must both be defined together, as lists')
-
-        if (cols and types):
-            validate_types(types)
-            if len(cols) != len(types):
-                logging.error('cols and types must be lists of equal length')
-            for i in len(cols):
-                series_res[col[i]] = ops[types[i]['type']](types[i], length)
-
-        else:
-            if not coltypes:
-                logging.error('please define either cols and types or coltypes')
-            # Assure iteritems compatibility throught 2.7 and 3+
-            try:
-                coltypes_items = coltypes.iteritems()
-            except AttributeError:
-                coltypes_items = coltypes.items()
-            for col, typ in coltypes_items:
-                data_builder = ops[typ['type']]
-                del typ['type']
-                series_res[col] = data_builder(typ, length)
+import logging
+import pandas as pd
 
-        return pd.DataFrame(series_res)
+from simulacrum import types
+
+TYPE_FUNCTIONS = {
+    'num': types.num_data,
+    'int': types.num_int,
+    'norm': types.norm_data,
+    'exp': types.exp_data,
+    'bin': types.binom_data,
+    'pois': types.poisson_data,
+    'txt': types.text_data,
+    'name': types.name_data,
+    'addr': types.address_data,
+    'date': types.date_data,
+    'coords': types.coords_data,
+    'uuid': types.uuid_data,
+    'faker': types.faker_data}
+
+
+def create(length=100, cols=None, types=None, coltypes=None):
+    """Create a dataset based on passed in information.
+
+    A user must either pass in cols and types lists, OR coltypes, OR the
+    function will default to "one of each" type.
+
+    Parameters
+    ----------
+    length : int, optional
+        How many records (rows) in the returned dataframe
+    cols : list, optional
+        A list of column names
+    types : list of dict, optional
+        A list of "type dictionaries", defining both the type, and any
+        parameters to be passed to the type function.
+    coltypes : dict, optional
+        A combined version of cols and types, where the keys of the dictionary
+        are cols, and the values are type dictionaries.
+    """
+    if cols and types and coltypes:
+        raise ValueError(
+            'coltypes should not be defined when cols and types are defined')
+    elif cols is not None and types is not None:
+        column_iter = cols
+        type_iter = types
+        if len(column_iter) != len(type_iter):
+            raise ValueError(
+                'cols and types must be lists of equal length')
+    else:
+        if coltypes is None:
+            logging.warning('No coltypes specified, using default')
+            coltypes = default_coltypes()
+        column_iter = coltypes.keys()
+        type_iter = coltypes.values()
+
+    series_res = {}
+    for col, type_dict in zip(column_iter, type_iter):
+        validate_type_dict(type_dict)
```

Comment:
I think it could be another PR.

A dictionnary of handlers like that could help:

````python
CHECK_TYPES_HANDLER = {
    "my_databuilder": a checktype function to use directly or another dictionnary containing each parameter type to check.
}

````